ballerina/subscription.bal

```diff
@@ -17,10 +17,140 @@
+        string responsePayload = check subscriberResponse.getTextPayload();
+        if challenge != responsePayload {
+            return;
+        }
+
+        VerifiedSubscription verifiedSubscription = {
+            ...message
+        };
+        check self.adaptor.callOnSubscriptionIntentVerifiedMethod(verifiedSubscription, headers);
+    }
+
+    isolated function validateSubscription(Subscription message, http:Headers headers) returns error? {
+        if self.isOnSubscriptionValidationAvailable {
+            return self.adaptor.callOnSubscriptionValidationMethod(message, headers);
+        }
+
+        if !message.hubCallback.startsWith("http://") && !message.hubCallback.startsWith("https://") {
+            return error SubscriptionDeniedError(
+                "Invalid hub.callback param in the request.", statusCode = http:STATUS_NOT_ACCEPTABLE);
+        }
+    }
+
+    isolated function initiateUnsubscription(Unsubscription message, http:Headers headers) returns http:Response {
+        if !self.isOnUnsubscriptionAvailable {
+            http:Response response = new;
+            response.statusCode = http:STATUS_ACCEPTED;
+            return response;
+        }
+
+        UnsubscriptionAccepted|error result = self.adaptor.callOnUnsubscriptionMethod(message, headers);
+        return processOnUnsubscriptionResult(result);
+    }
+
+    isolated function verifyUnsubscription(Unsubscription message, http:Headers headers) returns error? {
+        error? validationResult = self.validateUnsubscription(message, headers);
+        if validationResult is error {
+            [string, string?][] params = [
+                [HUB_MODE, MODE_DENIED],
+                [HUB_TOPIC, message.hubTopic],
+                [HUB_REASON, validationResult.message()]
+            ];
+            _ = check sendNotification(message.hubCallback, params, self.clientConfig);
+        }
+
+        string challenge = uuid:createType4AsString();
+        [string, string?][] params = [
+            [HUB_MODE, MODE_UNSUBSCRIBE],
+            [HUB_TOPIC, message.hubTopic],
+            [HUB_CHALLENGE, challenge]
+        ];
+        http:Response subscriberResponse = check sendNotification(message.hubCallback, params, self.clientConfig);
+        string responsePayload = check subscriberResponse.getTextPayload();
+        if challenge != responsePayload {
+            return;
+        }
+
+        VerifiedUnsubscription verifiedUnsubscription = {
+            ...message
+        };
+        check self.adaptor.callOnUnsubscriptionIntentVerifiedMethod(verifiedUnsubscription, headers);
+    }
+
+    isolated function validateUnsubscription(Unsubscription message, http:Headers headers) returns error? {
+        if self.isOnUnsubscriptionValidationAvailable {
+            return self.adaptor.callOnUnsubscriptionValidationMethod(message, headers);
+        }
+
+        if !message.hubCallback.startsWith("http://") && !message.hubCallback.startsWith("https://") {
```

Comment:
Should this be `or` instead of the `and` operator here?

Reply:
Logic here is if the callback did not start with `http://` and `https://` it should give an error right ? If we put a logica-or then imagine a scenario where the callback starts with `https://`, then the first condition will be true and it will return an error.

Reply:
Ah yes. Got it